contracts/utils/Swap.sol

```diff
@@ -0,0 +1,454 @@
+     */
+    function setFee(address referrer, uint256 fee) external onlyOwner {
+        require(referrer != address(0), "Referrer cannot be the zero address");
+        require(fee < MAX_PERCENTAGE, "Fee exceeds maximum");
+
+        referralFees[referrer] = fee;
+    }
+
+    /************************************************
+     *  OFFER CREATION AND SETTLEMENT
+     ***********************************************/
+
+    /**
+     * @notice Create a new offer available for swap
+     * @param offeredToken token offered by seller
+     * @param biddingToken token asked by seller
+     * @param minPrice minimum price of offeredToken in terms of biddingToken
+     * @param minBidSize minimum size allowed in terms of biddingToken
+     * @param totalSize amount of offeredToken offered by seller
+     */
+    function createOffer(
+        address offeredToken,
+        address biddingToken,
+        uint96 minPrice,
+        uint96 minBidSize,
+        uint128 totalSize
+    ) external override returns (uint256 swapId) {
+        require(
+            offeredToken != address(0),
+            "OfferedToken cannot be the zero address"
+        );
+        require(
+            biddingToken != address(0),
+            "BiddingToken cannot be the zero address"
+        );
+        require(minPrice > 0, "MinPrice must be larger than zero");
+        require(minBidSize > 0, "MinBidSize must be larger than zero");
+        require(totalSize > 0, "TotalSize must be larger than zero");
+
+        offersCounter += 1;
+
+        swapId = offersCounter;
+
+        swapOffers[swapId].seller = msg.sender;
+        swapOffers[swapId].isOpen = true;
+        swapOffers[swapId].offeredToken = offeredToken;
+        swapOffers[swapId].biddingToken = biddingToken;
+        swapOffers[swapId].offeredTokenDecimals = IERC20Detailed(offeredToken)
+            .decimals();
+        swapOffers[swapId].minBidSize = minBidSize;
+        swapOffers[swapId].minPrice = minPrice;
+        swapOffers[swapId].totalSize = totalSize;
+        swapOffers[swapId].availableSize = totalSize;
+
+        emit NewOffer(
+            swapId,
+            msg.sender,
+            offeredToken,
+            biddingToken,
+            minPrice,
+            minBidSize,
+            totalSize
+        );
+    }
+
+    /**
+     * @notice Settles the swap offering by iterating through the bids
+     * @param swapId unique identifier of the swap offer
+     * @param bids bids for swaps
+     */
+    function settleOffer(uint256 swapId, Bid[] calldata bids)
+        external
+        override
+        nonReentrant
+    {
+        Offer storage offer = swapOffers[swapId];
+
+        require(offer.seller != address(0), "Offer does not exist");
+        require(msg.sender == offer.seller, "Only seller can settle");
+        require(offer.isOpen, "Offer already closed");
+
+        for (uint256 i = 0; i < bids.length; i++) {
+            _swap(offer, bids[i]);
```

Comment:
We can cache the `offer.seller` here because it is repeated for all the swaps

```solidity
address seller = offer.seller
for (uint256 i = 0; i < bids.length; i++) {
            _swap(offer, seller, bids[i]);
}
```